### PR TITLE
chore(auth): load and store ASF device key from credential store

### PR DESCRIPTION
**Pull request by div5yesh**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
- update migration action
- remove codegen dep. from context data provider
- namespace device metadata store key with config params

*How did you test these changes?*
(Please add a line here how the changes were tested)

- [ ] Added Unit Tests
- [ ] Added Integration Tests

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt [hunk 1]
```diff
@@ -82,7 +82,7 @@ class AWSCognitoAuthPlugin : AuthPlugin<AWSCognitoAuthServiceBehavior>() {
                 configuration,
                 AWSCognitoAuthServiceBehavior.fromConfiguration(configuration),
                 credentialStoreClient,
-                configuration.userPool?.let { UserContextDataProvider(context, it) },
+                configuration.userPool?.let { UserContextDataProvider(context, it.poolId!!, it.appClient!!) },
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt:85]: Any way to safe-unwrap here?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 2]
```diff
@@ -192,7 +193,9 @@ internal class RealAWSCognitoAuthPlugin(
                 }
             }
 
-            val encodedContextData = authEnvironment.userContextDataProvider?.getEncodedContextData(username)
+            val asfDevice = authEnvironment.credentialStoreClient.loadCredentials(CredentialType.ASF)
+            val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:197]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 3]
```diff
@@ -275,7 +278,9 @@ internal class RealAWSCognitoAuthPlugin(
     ) {
         logger.verbose("ConfirmSignUp Starting execution")
         try {
-            val encodedContextData = authEnvironment.userContextDataProvider?.getEncodedContextData(username)
+            val asfDevice = authEnvironment.credentialStoreClient.loadCredentials(CredentialType.ASF)
+            val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:282]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 4]
```diff
@@ -339,7 +344,9 @@ internal class RealAWSCognitoAuthPlugin(
         logger.verbose("ResendSignUpCode Starting execution")
         try {
             val metadata = (options as? AWSCognitoAuthResendSignUpCodeOptions)?.metadata
-            val encodedContextData = authEnvironment.userContextDataProvider?.getEncodedContextData(username)
+            val asfDevice = authEnvironment.credentialStoreClient.loadCredentials(CredentialType.ASF)
+            val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:348]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 5]
```diff
@@ -970,9 +977,11 @@ internal class RealAWSCognitoAuthPlugin(
             )
 
             val appClient = requireNotNull(configuration.userPool?.appClient)
-            val encodedData = authEnvironment.userContextDataProvider?.getEncodedContextData(username)
-
             GlobalScope.launch {
+                val asfDevice = authEnvironment.credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:982]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 6]
```diff
@@ -1015,7 +1024,12 @@ internal class RealAWSCognitoAuthPlugin(
 
             GlobalScope.launch {
                 try {
-                    val encodedContextData = authEnvironment.userContextDataProvider?.getEncodedContextData(username)
+                    val asfDevice = authEnvironment.credentialStoreClient.loadCredentials(CredentialType.ASF)
+                    val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:1028]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/DeviceSRPCognitoSignInActions.kt [hunk 7]
```diff
@@ -52,7 +52,9 @@ internal object DeviceSRPCognitoSignInActions : DeviceSRPSignInActions {
             logger.verbose("$id Starting execution")
             val username = event.username
             val evt = try {
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(username)
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/DeviceSRPCognitoSignInActions.kt:56]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/DeviceSRPCognitoSignInActions.kt [hunk 8]
```diff
@@ -118,7 +120,10 @@ internal object DeviceSRPCognitoSignInActions : DeviceSRPSignInActions {
                 val srpB = params.getValue(KEY_SRP_B)
                 val deviceKey = params.getValue(KEY_DEVICE_KEY)
                 val deviceGroupKey = params.getValue(KEY_DEVICE_GROUP_KEY)
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/DeviceSRPCognitoSignInActions.kt:125]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/FetchAuthSessionCognitoActions.kt [hunk 9]
```diff
@@ -62,7 +62,10 @@ internal object FetchAuthSessionCognitoActions : FetchAuthSessionActions {
                 )
                 tokens.refreshToken?.let { authParameters[KEY_REFRESH_TOKEN] = it }
                 secretHash?.let { authParameters[KEY_SECRET_HASH] = it }
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/FetchAuthSessionCognitoActions.kt:67]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/MigrateAuthCognitoActions.kt [hunk 10]
```diff
@@ -44,7 +46,10 @@ internal object MigrateAuthCognitoActions : MigrateAuthActions {
 
                 val authParams = mutableMapOf(KEY_USERNAME to event.username, KEY_PASSWORD to event.password)
                 secretHash?.let { authParams[KEY_SECRET_HASH] = it }
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(event.username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/MigrateAuthCognitoActions.kt:51]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SRPCognitoActions.kt [hunk 11]
```diff
@@ -62,9 +62,11 @@ internal object SRPCognitoActions : SRPActions {
                 )
 
                 val authParams = mutableMapOf(KEY_USERNAME to event.username, KEY_SRP_A to srpHelper.getPublicA())
-
                 secretHash?.let { authParams[KEY_SECRET_HASH] = it }
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(event.username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SRPCognitoActions.kt:68]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SRPCognitoActions.kt [hunk 12]
```diff
@@ -117,7 +119,10 @@ internal object SRPCognitoActions : SRPActions {
                     KEY_CHALLENGE_NAME to VALUE_SRP_A
                 )
                 secretHash?.let { authParams[KEY_SECRET_HASH] = it }
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(event.username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SRPCognitoActions.kt:124]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SRPCognitoActions.kt [hunk 13]
```diff
@@ -185,7 +190,10 @@ internal object SRPCognitoActions : SRPActions {
                     KEY_TIMESTAMP to srpHelper.dateString,
                 )
                 secretHash?.let { challengeParams[KEY_SECRET_HASH] = it }
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SRPCognitoActions.kt:195]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SignInChallengeCognitoActions.kt [hunk 14]
```diff
@@ -57,7 +59,10 @@ internal object SignInChallengeCognitoActions : SignInChallengeActions {
                 configuration.userPool?.appClientSecret
             )
             secretHash?.let { challengeResponses[KEY_SECRET_HASH] = it }
-            val encodedContextData = username?.let { userContextDataProvider?.getEncodedContextData(it) }
+
+            val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+            val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SignInChallengeCognitoActions.kt:64]: safe cast preferred (as?)

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SignInCustomCognitoActions.kt [hunk 15]
```diff
@@ -48,7 +48,10 @@ internal object SignInCustomCognitoActions : CustomSignInActions {
 
                 val authParams = mutableMapOf(KEY_USERNAME to event.username)
                 secretHash?.let { authParams[KEY_SECRET_HASH] = it }
-                val encodedContextData = userContextDataProvider?.getEncodedContextData(event.username)
+
+                val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
+                val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SignInCustomCognitoActions.kt:53]: safe cast preferred (as?)

**Review by tylerjroach — CHANGES_REQUESTED** [review]
```
val asfDevice = credentialStoreClient.loadCredentials(CredentialType.ASF)
val deviceId = (asfDevice as AmplifyCredential.ASFDevice).id
val encodedContextData = userContextDataProvider?.getEncodedContextData(event.username, deviceId)
```

These 3 lines with unsafe case are used 10+ times. Can we add a helper method to get encodedContextData, as well as improving with safe casting.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt (hunk 1 above)
Comment (div5yesh) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt:85]: I have a follow up PR for refactoring all the `!!` and `?: ""`.

**Review by gpanshu — APPROVED** [review]

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**